### PR TITLE
Add caller sdk id and version to telemetry server side and client side

### LINKED ISSUE #4863: [Feature Request] Caller SDK details in ExtraQueryParams
### Problem statement

# Goal

Higher level SDKs provide a hint on their identity: id + version

# Dev Experience

**request**.WithExtraQueryParameters({{ "caller-sdk-id", "41"}, { "caller-sdk-ver", "1.42.0"}}

Note: pls improve WithExtraQueryParameters so that it can be used multiple times, in which case values are appended.

# Implementation

The data should be sent to ESTS via the msal-curr header and also captured in the OTEL metrics.

### PR DESCRIPTION
**Pull request by neha-bhargava**
Fixes #4863 

**Changes proposed in this request**
Add caller sdk details to client side and server side telemetry

**Testing**
Unit and integration tests

**Performance impact**

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs [hunk 1]
```diff
@@ -21,7 +21,10 @@ public static class AcquireTokenParameterBuilderExtensions
             IDictionary<string, string> extraHttpHeaders)
             where T : AbstractAcquireTokenParameterBuilder<T>
         {
-            builder.CommonParameters.ExtraHttpHeaders = extraHttpHeaders;
+            foreach (var kvp in extraHttpHeaders)
+            {
+                builder.CommonParameters.ExtraHttpHeaders.Add(kvp.Key, kvp.Value);
```
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs:26]: This change addresses the statement in the issue: `Note: pls improve WithExtraQueryParameters so that it can be used multiple times, in which case values are appended.`

So the headers can be added multiple times in the flow and will be maintained. 

The `ExtraHttpHeaders` here are later added to the headers sent to ests [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs#L219). Do you mean that the headers that MSAL sends to ests like `x-app-name` can get overridden accidentally?

[the remark this answers is not in the document]
Comment (localden) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs:26]: @neha-bhargava yep, exactly! I was curious if someone can inadvertently replace the headers that we need for eSTS telemetry.

[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs:26]: That is a possibility currently, I think we should add these first and then add the common headers (the one MSAL sends to ests) so that the final set have all the expected headers.

File: src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs [hunk 2]
```diff
@@ -242,6 +244,18 @@ private ApiEvent InitializeApiEvent(string accountId)
             apiEvent.TokenType = AuthenticationRequestParameters.AuthenticationScheme.TelemetryTokenType;
             apiEvent.AssertionType = GetAssertionType();
 
+            if (AuthenticationRequestParameters.ExtraQueryParameters.ContainsKey("caller-sdk-id"))
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:247]: nit: You can use the `TryGet(key, out string val)` to combine ContainsKey and the [key] operations toghether.

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs [hunk 3]
```diff
@@ -135,7 +135,11 @@ public string TokenTypeString
         public CacheLevel CacheLevel { get; set; }
 
         public string MsalRuntimeTelemetry { get; set; }
-        
+
+        public string CallerSdkApiId { get; set; } = "";
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs:139]: why not let them be null?

**Review by bgavrilMS — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs [hunk 4]
```diff
@@ -21,7 +21,10 @@ public static class AcquireTokenParameterBuilderExtensions
             IDictionary<string, string> extraHttpHeaders)
             where T : AbstractAcquireTokenParameterBuilder<T>
         {
-            builder.CommonParameters.ExtraHttpHeaders = extraHttpHeaders;
+            foreach (var kvp in extraHttpHeaders)
```
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs:24]: Re-read the email. Seems like its just ExtraParameters mentioned in the email. I think that sounds better since these are neither headers nor query parameter. I can add another API in this extension class `.WithExtraParameters`

[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs:24]: CC: @jmprieur @localden

[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs:24]: We haven't really decided on WithExtraParameters yet. Still discussing with @jmprieur and @localden  if this is the right approach for all 1p features.

File: src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenParameterBuilderExtensions.cs:26]: Reverting this since we will use extraQueryParameters

File: src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs [hunk 5]
```diff
@@ -242,6 +244,19 @@ private ApiEvent InitializeApiEvent(string accountId)
             apiEvent.TokenType = AuthenticationRequestParameters.AuthenticationScheme.TelemetryTokenType;
             apiEvent.AssertionType = GetAssertionType();
 
+            // Check if ExtraQueryParameters contains caller-sdk-id and caller-sdk-ver
+            if (AuthenticationRequestParameters.ExtraQueryParameters.TryGetValue("caller-sdk-id", out string callerSdkId))
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:248]: Pls keep the "Extra query params" way, in addition to `WithClientName` and `WithClientVersion`. Some SDKs like Id.Web create just 1 CCA object, for use with various APIs. So we need the ability of configuring "caller SDK" on a per request basis. 

CC @localden @rayluo

File: tests/Microsoft.Identity.Test.Unit/AppConfigTests/ConfidentialClientApplicationBuilderTests.cs [hunk 6]
```diff
@@ -41,8 +41,6 @@ public void TestConstructor()
             // Validate Defaults
             Assert.AreEqual(LogLevel.Info, cca.AppConfig.LogLevel);
             Assert.AreEqual(TestConstants.ClientId, cca.AppConfig.ClientId);
-            Assert.IsNotNull(cca.AppConfig.ClientName);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/AppConfigTests/ConfidentialClientApplicationBuilderTests.cs:44]: What are they if not null?

Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Unit/AppConfigTests/ConfidentialClientApplicationBuilderTests.cs:44]: Since we are changing this to accept the api id of the caller sdk, these can be null or empty when not passed in by the user. I can add the check to validate that these are null by default. And have a value when a value is passed in

File: src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs [hunk 7]
```diff
@@ -37,24 +37,21 @@ public ApplicationConfiguration(MsalClientType applicationType)
...
-        public const string DefaultClientVersion = "0.0.0.0";
 
         // For telemetry, the ClientName of the application.
-        private string _clientName = DefaultClientName;
+        private string _clientName;
         public string ClientName
         {
             get => _clientName;
-            internal set { _clientName = string.IsNullOrWhiteSpace(value) ? DefaultClientName : value; }
+            internal set { _clientName = string.IsNullOrWhiteSpace(value) ? "" : value; }
```
Comment (TimHannMSFT) [inline src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs:46]:
```suggestion
            internal set { _clientName = string.IsNullOrWhiteSpace(value) ? string.Empty : value; }

```

File: src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs [hunk 8]
```diff
@@ -37,24 +37,21 @@ public ApplicationConfiguration(MsalClientType applicationType)
...
         }
 
         // For telemetry, the ClientVersion of the application.
-        private string _clientVersion = DefaultClientVersion;
+        private string _clientVersion;
         public string ClientVersion
         {
             get => _clientVersion;
-            internal set { _clientVersion = string.IsNullOrWhiteSpace(value) ? DefaultClientVersion : value; }
+            internal set { _clientVersion = string.IsNullOrWhiteSpace(value) ? "" : value; }
```
Comment (TimHannMSFT) [inline src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs:54]:
```suggestion
            internal set { _clientVersion = string.IsNullOrWhiteSpace(value) ? string.Empty : value; }

```

File: src/client/Microsoft.Identity.Client/Platforms/Features/OpenTelemetry/OtelInstrumentation.cs [hunk 9]
```diff
@@ -162,6 +168,8 @@ public void IncrementSuccessCounter(string platform,
                         new(TelemetryConstants.MsalVersion, MsalIdHelper.GetMsalVersion()),
                         new(TelemetryConstants.Platform, platform),
                         new(TelemetryConstants.ApiId, apiId),
+                        new(TelemetryConstants.CallerSdkId, callerSdkId ?? ""),
```
Comment (TimHannMSFT) [inline src/client/Microsoft.Identity.Client/Platforms/Features/OpenTelemetry/OtelInstrumentation.cs:171]:
```suggestion
                        new(TelemetryConstants.CallerSdkId, callerSdkId ?? string.Empty),
```

File: src/client/Microsoft.Identity.Client/Platforms/Features/OpenTelemetry/OtelInstrumentation.cs [hunk 10]
```diff
@@ -162,6 +168,8 @@ public void IncrementSuccessCounter(string platform,
...
+                        new(TelemetryConstants.CallerSdkVersion, callerSdkVersion ?? ""),
```
Comment (TimHannMSFT) [inline src/client/Microsoft.Identity.Client/Platforms/Features/OpenTelemetry/OtelInstrumentation.cs:172]:
```suggestion
                        new(TelemetryConstants.CallerSdkVersion, callerSdkVersion ?? string.Empty),
```

Consider using something that will be more readable when reviewing though as well e.g. "Unknown"

Is this some value a caller optionally sets up when building out the CCA/PCA? Is there an update to our internal libraries which leverage this?

File: src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs [hunk 11]
```diff
@@ -51,6 +51,10 @@ public string GetCurrentRequestHeader(ApiEvent eventInProgress)
             sb.Append(TelemetryConstants.CommaDelimiter);
             // Token type is to indicate 1 - bearer, 2 - pop, 3 - ssh-cert, 4 - external.
             sb.Append(eventInProgress.TokenTypeString);
+            sb.Append(TelemetryConstants.CommaDelimiter);
```
Comment (TimHannMSFT) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs:54]: Should this be wrapped in a null check for these values? Or is it okay to have the string with the empty sections? e.g. ",1,,"?

Who is parsing this data out? How are they using it?

File: src/client/Microsoft.Identity.Client/TelemetryCore/OpenTelemetry/IOtelInstrumentation.cs [hunk 12]
```diff
@@ -14,13 +14,17 @@ internal interface IOtelInstrumentation
...
             AuthenticationResultMetadata authResultMetadata,
             ILoggerAdapter logger);
 
         internal void IncrementSuccessCounter(string platform,
             ApiEvent.ApiIds apiId,
+            string callerSdkId,
+            string callerSdkVersion,
             TokenSource tokenSource,
             CacheRefreshReason cacheRefreshReason,
             CacheLevel cacheLevel,
```
Comment (TimHannMSFT) [inline src/client/Microsoft.Identity.Client/TelemetryCore/OpenTelemetry/IOtelInstrumentation.cs:30]: Why not add the caller SDK/version on the log failure below as well? What if for some reason 100% of calls fail with a new version?

**Comment by TimHannMSFT** [comment]
Thanks @neha-bhargava! Can you help me understand how these fields are meant to interact with the `supplemental-version-info` currently getting sent in client telemetry we discussed?

As we discussed, please be sure to update the schema here once this is finalized: https://eng.ms/docs/microsoft-security/identity/entra-developer-application-platform/id4s-identity-for-services/authn-middleware-sdk-microsoft-identity-service-essentials/microsoft-identity-service-essentials/articles/using-mise/telemetry-schema

File: src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs (hunk 11 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs:54]: We parse this data and when a value is not available we expect the empty spaces. Data is parsed as a sequence.

**Comment by neha-bhargava** [comment]
> Thanks @neha-bhargava! Can you help me understand how these fields are meant to interact with the `supplemental-version-info` currently getting sent in client telemetry we discussed?
> 
> As we discussed, please be sure to update the schema here once this is finalized: https://eng.ms/docs/microsoft-security/identity/entra-developer-application-platform/id4s-identity-for-services/authn-middleware-sdk-microsoft-identity-service-essentials/microsoft-identity-service-essentials/articles/using-mise/telemetry-schema

Since this value is to recognize the API used, this value can be added per request too. So, the supplemental-version-info could not be used in this case. Also we need to send this information to server side telemetry hence using this API.

File: src/client/Microsoft.Identity.Client/AppConfig/BaseAbstractApplicationBuilder.cs [hunk 13]
```diff
@@ -204,6 +204,28 @@ public T WithExperimentalFeatures(bool enableExperimentalFeatures = true)
             return (T)this;
         }
 
+        /// <summary>
+        /// Sets the name of the calling SDK API Id for telemetry purposes.
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/AppConfig/BaseAbstractApplicationBuilder.cs:208]:
```suggestion
        /// Sets the name of the calling SDK API for telemetry purposes.
```

File: src/client/Microsoft.Identity.Client/AppConfig/BaseAbstractApplicationBuilder.cs [hunk 14]
```diff
@@ -204,6 +204,28 @@ public T WithExperimentalFeatures(bool enableExperimentalFeatures = true)
...
+        /// </summary>
+        /// <param name="clientName">The name of the SDK API Id for telemetry purposes.</param>
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/AppConfig/BaseAbstractApplicationBuilder.cs:210]:
```suggestion
        /// <param name="clientName">The name of the SDK API for telemetry purposes.</param>
```

File: src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs [hunk 15]
```diff
@@ -242,12 +247,42 @@ private ApiEvent InitializeApiEvent(string accountId)
...
+
             // Give derived classes the ability to add or modify fields in the telemetry as needed.
             EnrichTelemetryApiEvent(apiEvent);
 
             return apiEvent;
         }
 
+        private void UpdateCallerSdkDetails(ApiEvent apiEvent)
+        {
+            string callerSdkId;
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:260]: is this an id or the SDK name?

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryConstants.cs [hunk 16]
```diff
@@ -44,6 +44,8 @@ internal static class TelemetryConstants
         public const string Platform = "Platform";
         public const string ApiId = "ApiId";
         public const string IsProactiveRefresh = "IsProactiveRefresh";
+        public const string CallerSdkId = "CallerSdkId";
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryConstants.cs:47]: Id or name?

File: tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs [hunk 17]
```diff
@@ -390,6 +361,87 @@ await pca.AcquireTokenByUsernamePassword(TestConstants.s_scope, "username", Test
...
+                    .WithExtraQueryParameters(new Dictionary<string, string> {
+                        { "caller-sdk-id", callerSdkId },
+                        { "caller-sdk-ver", callerSdkVersion } })
+                    .ExecuteAsync().ConfigureAwait(false);
+
+                AssertCurrentTelemetry(
+                    requestHandler.ActualRequestMessage,
+                    ApiIds.AcquireTokenForClient,
+                    CacheRefreshReason.NoCachedAccessToken,
+                    callerSdkId: callerSdkId.Substring(0, Math.Min(callerSdkId.Length, 10)),
```
Comment (gladjohn) [inline tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs:412]: I would move these numbers 10 and 20 to constants and add a meaningful name to better understand the test

File: tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs [hunk 18]
```diff
@@ -390,6 +361,87 @@ await pca.AcquireTokenByUsernamePassword(TestConstants.s_scope, "username", Test
...
+                    requestHandler.ActualRequestMessage, 
+                    ApiIds.AcquireTokenForClient, 
+                    CacheRefreshReason.NoCachedAccessToken,
+                    callerSdkId: "testApiId", 
+                    callerSdkVersion: "testSdkVersion");
+            }
+        }
+
+        [TestMethod]
+        public async Task CallerSdkDetails_ConstraintsTestAsync()
```
Comment (gladjohn) [inline tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs:390]: do we have a ` CallerSdkDetails_NullValuesTestAsync` test to check for null id and versionb?

Comment (gladjohn) [inline tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs:390]: you can also add some edge cases, like a test for `CallerSdkDetails_MaximumLengthTestAsync` or make this a parameterized test with different lengths for id and name

File: tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs [hunk 19]
```diff
@@ -390,6 +361,87 @@ await pca.AcquireTokenByUsernamePassword(TestConstants.s_scope, "username", Test
...
+        {
+            string callerSdkId = "testApiIdMoreThan10Chars";
```
Comment (gladjohn) [inline tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs:392]: you should test invalid inputs like special characters or symbols and ensure the system handles them appropriately

**Review by gladjohn — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs (hunk 15 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:260]: No Api id of the api used by the caller SDK

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryConstants.cs (hunk 16 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryConstants.cs:47]: It will be an id like `1001` or something.

File: tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs (hunk 19 above)
Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs:392]: I don't think we need tests to that extent. This is for internal use. Mostly used by Id web and Azure Identity

File: tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs (hunk 18 above)
Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Unit/TelemetryTests/HttpTelemetryTests.cs:390]: IMO this one test should be enough along with the test we already have for ids within the range and null/ empty values which are already there.
